### PR TITLE
Use Rust FFI #2: Offload installers

### PR DESCRIPTION
**Pull request by tienvx**
This is 2nd pull request in a series of "small" pull requests I split from https://github.com/pact-foundation/pact-php/pull/210

This pull request depend on https://github.com/pact-foundation/pact-php/pull/278, please review it first.

There are number of reasons why I choose this approach:
* Update installers to downloads more files require adding quite a mount of code (see pull request above)
  * Need download at least 4 files
    * Pact Ruby Standalone
    * Pact FFI header
    * Pact FFI shared library
    * Pact stub server
  *  If we want supports plugins, 2 more for each plugin
      * csv
          * json meta file
          * executable file
      * protobuf
          * json meta file
          * executable file
  * Each file has different url/file-name/destination-path depend on OS and architecture.
  * Extracting `zip` or `tar.gz` file is easy, but extracting `gzip` file require more work, and we need to move the target file to the correct destination. `*.gz` don't restore the original file name.
  * Mark files as executable after downloading
  * Delete downloaded archived file after extracting
  * Determine correct path to scripts for each OS
* Maintain custom code does not look right to me
  * Download on runtime, not on running `composer install/update`
  * Need to re-download all files after installing/updating
  * No caching support

After looking for alternate solution, I found this library [civicrm/composer-downloads-plugin](https://github.com/civicrm/composer-downloads-plugin). But the author is also busy, he don't have time review my pull requests.

That's why I maintain my own fork. After making some changes, I released version 1.1, which is used by this pull request. This library has some features:
* Support all kind of files that composer suppport
* Download on `composer install/update` command
* Support caching from composer
* Don't need to re-download files if already downloaded
* Support variables for different url/path depend on OS and architecture.

I think this pull request also solve this ticket https://github.com/pact-foundation/pact-php/issues/276

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -69,13 +70,29 @@
...
+                },
+                "url": "https://github.com/pact-foundation/pact-ruby-standalone/releases/download/v{$version}/pact-{$version}-{$os}{$architecture}.{$extension}",
+                "path": "bin/pact-ruby-standalone"
+            }
+        }
+    },
+    "config": {
+        "allow-plugins": {
+            "tienvx/composer-downloads-plugin": true
+        }
```
Comment (tienvx) [inline composer.json:96]: Add this config to make running test easier. I think each project depend on this library also need to define this config again in `composer.json`

File: src/PhpPact/Standalone/Installer/Model/Scripts.php [hunk 2]
```diff
@@ -4,101 +4,59 @@
...
     }
 
-    /**
-     * @return string
-     */
-    public function getPactMessage(): string
+    private static function getSuffix(): string
     {
-        return $this->pactMessage;
+        return (PHP_OS_FAMILY === 'Windows' ? '.bat' : '');
```
Comment (tienvx) [inline src/PhpPact/Standalone/Installer/Model/Scripts.php:60]: If we use `symfony/process`, I think we don't have to do this.

We can discuss this again in the future pull request (maybe the 4th pull request)

File: script/create-pr-to-update-pact-ruby-standalone.sh [hunk 3]
```diff
@@ -1,31 +0,0 @@
...
-done
-
-
-git add ${VERSION_FILES}
-git commit -m "feat(upgrade): update standalone to ${STANDALONE_VERSION}"
-git push --set-upstream origin ${BRANCH_NAME}
-
-hub pull-request --browse --message "feat: update standalone to ${STANDALONE_VERSION}"
-
-git checkout master
```
Comment (tienvx) [inline script/create-pr-to-update-pact-ruby-standalone.sh:31]: Now we only need to update 1 number in `composer.json`. That's why I think this script is not needed.

File: src/PhpPact/Standalone/Installer/Model/Scripts.php (hunk 2 above)
Comment (tienvx) [reply src/PhpPact/Standalone/Installer/Model/Scripts.php:60]: https://github.com/pact-foundation/pact-php/pull/284

**Review by cfmack — APPROVED** [review]
Code looks good, I ran this locally and it passed, all tests in Github actions pass.

File: composer.json (hunk 1 above)
Comment (YOU54F) [reply composer.json:96]: I think if we can provide support for a user to support the ffi being provided by them in their own location (which should be fine as we can have to provide a path to the library loader) we can support users who dont want to, or cant enable scripts to download.

Comment (tienvx) [reply composer.json:96]: I agree with it.

This feature can be useful for plugins. See this [link](https://pact-foundation.slack.com/archives/C047TCR7B6W/p1679881619953729) for explanation.

Currently I don't have any idea on how to do it. Can you create a ticket so we can look at it later?

Comment (YOU54F) [reply composer.json:96]: of course buddy 👍🏾  shall do now

**Pull request merged**